web/api/v1/api_test.go

```diff
@@ -55,55 +56,98 @@ import (
 	"github.com/prometheus/prometheus/util/testutil"
 )
 
-type testTargetRetriever struct{}
+// testMetaStore satisfies the scrape.MetricMetadataStore interface.
+// It is used to inject specific metadata as part of a test case.
+type testMetaStore struct {
+	Metadata []scrape.MetricMetadata
+}
+
+func (s *testMetaStore) ListMetadata() []scrape.MetricMetadata {
+	return s.Metadata
+}
+
+func (s *testMetaStore) GetMetadata(metric string) (scrape.MetricMetadata, bool) {
+	for _, m := range s.Metadata {
+		if metric == m.Metric {
+			return m, true
+		}
+	}
+
+	return scrape.MetricMetadata{}, false
+}
+
+// testTargetRetriever represents a list of targets to scrape.
+// It is used to represent targets as part of test cases.
+type testTargetRetriever struct {
+	activeTargets  map[string][]*scrape.Target
+	droppedTargets map[string][]*scrape.Target
+}
+
+type testTargetParams struct {
+	Identifier       string
+	Labels           []labels.Label
+	DiscoveredLabels []labels.Label
+	Params           url.Values
+	Reports          []*testReport
+	Active           bool
+}
+
+type testReport struct {
+	Start    time.Time
+	Duration time.Duration
+	Error    error
+}
+
+func newTestTargetRetriever(targetsInfo []*testTargetParams) *testTargetRetriever {
+	var activeTargets map[string][]*scrape.Target
+	var droppedTargets map[string][]*scrape.Target
+	activeTargets = make(map[string][]*scrape.Target)
+	droppedTargets = make(map[string][]*scrape.Target)
+
+	for _, t := range targetsInfo {
+		nt := scrape.NewTarget(t.Labels, t.DiscoveredLabels, t.Params)
+
+		for _, r := range t.Reports {
+			nt.Report(r.Start, r.Duration, r.Error)
+		}
+
+		if t.Active {
+			activeTargets[t.Identifier] = []*scrape.Target{nt}
+		} else {
+			droppedTargets[t.Identifier] = []*scrape.Target{nt}
+		}
+	}
+
+	return &testTargetRetriever{
+		activeTargets:  activeTargets,
+		droppedTargets: droppedTargets,
+	}
+}
 
 var (
 	scrapeStart = time.Now().Add(-11 * time.Second)
 )
 
 func (t testTargetRetriever) TargetsActive() map[string][]*scrape.Target {
-	testTarget := scrape.NewTarget(
```

Comment:
While this change is technically not needed right now. It will in my subsequent pull request as I need to define various targets at different tests cases dynamically.

Would rather get some feedback on it now and avoid a bigger diff later on.